code/datums/alarm.dm

```diff
@@ -0,0 +1,199 @@
+//This files deals with the generic sending and receiving of "alarms"
+//This is a somewhat blanket term, it covers things like fire/power/atmos alarms, along with some oddballs
+//Side effect of how the system is used, these are mostly things that are of interest to ais and borgs
+//Though it could easily be expanded to cover other senders/revievers
+//The system as a whole differs from reading off a global list in a few ways.
+//In that A, it allows us to send cameras for ais/borgs/potentially others to jump to
+//And B, it's not like we're giving you all the alarms that have been sent, because of the seperate listing for each reviever
+//You only recieve alarms sent after you start to listen
+//Also of note, due to an optimzation done on areas, one alarm handler will only ever send one "on" or "off" alarm
+//So the whole only receving stuff sent post creation thing actually matters
+//Honestly I'm not sure how much of this is a feature, and how much is just old code
+//But I'm leaving it how I found it
+
+///Represents a single source of alarms, one alarm handler will only ever count for one alarm per listener
+/datum/alarm_handler
+	///A list of alarm type -> list of areas we currently have alarms in
+	var/list/sent_alarms = list()
+	///Our source atom
+	var/atom/source_atom
+
+/datum/alarm_handler/New(atom/source_atom)
+	if(istype(source_atom))
+		src.source_atom = source_atom
+	else
+		var/source_type = ""
+		if(istype(source_atom, /datum))
+			source_type = source_atom.type
+		stack_trace("a non atom was passed into alarm_handler! [source_atom] [source_type]")
```

Comment:
```suggestion
		stack_trace("a non atom was passed into alarm_handler! [source_atom] [istype(source_atom, /datum) ? source_atom.type : ""]")
```

Reply:
Or if you don't like how much this fits on one line, make the var declaration one line.

Reply:
I can do that